Bug 1931217: add affinity groups to oVirt installer

Gal-Zaidman opened the pull request:
We want the installer to have the following behavior:

Default:
Create 2 unique affinity groups with a hard negative affinity rule:
- ovirt11-****-controlplane - contain the masters only and have a higher priority(5).
- ovirt11-****-compute - contain the compute only, priority(3).

Customizations:
1. No Affinity groups:
To not create any AffinityGroups the user needs to have an empty field on the install config for the platform Like:
```yaml
...
platform:
  ovirt:
    api_vip: xxx.xxx.xxx.xxx
    ingress_vip: xxx.xxx.xxx.xxx
    ovirt_cluster_id: xxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxx
    ovirt_storage_domain_id: xxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxx
    affinityGroups: []
...
```

2. Customize created affinity Groups:
The user can define the affinity groups that the installer will create by editing the affinityGroups field, for example:
```yaml
platform:
  ovirt:
    api_vip: xxx.xxx.xxx.xxx
    ingress_vip: xxx.xxx.xxx.xxx
    ovirt_cluster_id: xxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxx
    ovirt_storage_domain_id: xxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxx
    affinityGroups:
    - description: test1
      enforcing: true
      name: test1
      priority: 3
    - description: test2
        host
      enforcing: true
      name: test2
      priority: 5
...
```

3. Decide which machine pools will be added to the affinity groups:
The user can specify on the machine pool the affinity groups that those machines will join on creation, by editing the affinityGroupsNames field, those groups need to be created by the installer or existing in the ovirt cluster.
```yaml
...
compute:
- name: worker
  platform:
    ovirt:
      affinityGroupsNames:
      - test1
...
  replicas: 2
controlPlane:
...
  name: master
  platform:
    ovirt:
      affinityGroupsNames:
      - test1
      - test2
...
  replicas: 3
...

Gal-Zaidman commented:
/hold

Need https://github.com/openshift/cluster-api-provider-ovirt/pull/80 to work

janoszen commented on data/data/ovirt/affinity_group/main.tf:
```diff
@@ -0,0 +1,18 @@
+  cluster_id   = var.ovirt_cluster_id
+  priority     = var.ovirt_control_plane_affinity_groups[count.index]["priority"]
+  vm_positive  = false
+  vm_enforcing = false
```
`vm_enforcing=false` means that the masters MAY end up on the same physical host if the user does not have enough hosts instead of failing. This can lead to inadvertently having a single point of failure.

This option should be customizable and I would strongly argue for setting the default value to `true` to stay in line with all Kubernetes deployment recommendations that I know of.

Gal-Zaidman replied:
Making this configurable and default to true

janoszen commented on data/data/ovirt/affinity_group/main.tf:
```diff
@@ -0,0 +1,18 @@
+  cluster_id   = var.ovirt_cluster_id
+  priority     = var.ovirt_compute_affinity_groups[count.index]["priority"]
+  vm_positive  = false
+  vm_enforcing = false
```
This option should be customizable. Furthermore, there should be a customizable description with the following text:

> AG for spreading the load out to multiple hosts without hard enforcement.

Gal-Zaidman replied:
+1

janoszen commented on pkg/tfvars/ovirt/ovirt.go:
```diff
@@ -71,6 +78,50 @@ func TFVars(
 	return json.MarshalIndent(cfg, "", "  ")
 }
+
+func createControlPlaneAffinityGroups(masterAffiniryGroups []*ovirt.AffinityGroup, infraID string) []map[string]interface{} {
```
Typo!

Gal-Zaidman replied:
done

janoszen commented on pkg/tfvars/ovirt/ovirt.go:
```diff
@@ -71,6 +78,50 @@ func TFVars(
+func createControlPlaneAffinityGroups(masterAffiniryGroups []*ovirt.AffinityGroup, infraID string) []map[string]interface{} {
+	var agMap []map[string]interface{}
+	for _, ag := range masterAffiniryGroups {
+		if ag.Name != "" {
```
Can we please fail if the user doesn't input a name for the AG?

staebler replied:
Failing here is good, but I would hope that this would have already been caught by validation.

Gal-Zaidman replied:
Currently, it is not, but I'll add that to the validations

janoszen commented on pkg/tfvars/ovirt/ovirt.go:
```diff
@@ -71,6 +78,50 @@ func TFVars(
+
+	for _, pool := range computeMachines {
+		for _, poolAG := range pool.AffinityGroups {
+			if poolAG.Name != "" {
```
Let's fail if the user doesn't enter an AG name.

Gal-Zaidman replied:
will be handled by validations

staebler commented on pkg/tfvars/ovirt/ovirt.go:
```diff
@@ -3,6 +3,9 @@ package ovirt
 
 import (
 	"encoding/json"
+	"github.com/openshift/installer/pkg/types/ovirt"
+	ovirtdefaults "github.com/openshift/installer/pkg/types/ovirt/defaults"
```
Wrong location for these imports.

Gal-Zaidman replied:
Done

staebler commented on pkg/tfvars/ovirt/ovirt.go:
```diff
@@ -71,6 +78,50 @@ func TFVars(
+
+func createComputeAffinityGroups(computeMachines []*ovirt.MachinePool, masterAffiniryGroups []*ovirt.AffinityGroup, infraID string) []map[string]interface{} {
+	var agMap []map[string]interface{}
+	uniqeAGName := make(map[string]*ovirt.AffinityGroup)
```
```suggestion
	uniqueAGName := make(map[string]*ovirt.AffinityGroup)
```

Gal-Zaidman replied:
done

staebler commented on pkg/tfvars/ovirt/ovirt.go:
```diff
@@ -71,6 +78,50 @@ func TFVars(
+			}
+		}
+	}
+	// If the affinity group exist on master list don't create it twice
```
What does the `vm_list` field of the `ovirt_affinity_group` terraform resource indicate? If an affinity group is included in both the masters and compute affinity groups in the install config, then the resulting affinity group will have a `vm_list` that only includes the masters.

Gal-Zaidman replied:
vm_list fields are VMs that will be added to the affinity group.
This field will only exists in the `control_plain_affinity_group` because the masters are created by terraform so machine API create logic will not be called for them.
In case a affinity group exist in the masters and workers then the masters will be added to it by terraform with the vm_list field and the workers will be added by the cluster API provider create logic

staebler replied:
Makes sense. Thanks for the explanation.

staebler commented on pkg/types/ovirt/defaults/machinepools.go:
```diff
@@ -0,0 +1,42 @@
+
+const (
+	//DAffinyGroupBase is a holder which will be replaced with the cluster name when creating the affinity group
+	DAffinyGroupBase                 = "CLUSTERNAME"
```
I strongly discourage using this as a means of coming up with an affinity group that includes the cluster name. Instead, I recommend adding the cluster name to the beginning or end of the name provided by the user. At a minimum, if you go the route of using a string to be replaced, then you need to document that around the type and add validation that enforces the existing of the string exactly once in the name.

janoszen replied:
@Gal-Zaidman any thoughts on this?

Gal-Zaidman replied:
Well currently we use the replace to add the cluster name at the beginning.
The reason I used the place holder is that if a user inputs a custom name then it will not be replaced and if the default configuration is applied then the string will be replaced.
Instead, I can add some if to check if the affinity group name is different then the default affinity group name...

staebler replied:
Oh, so you are *not* expecting the user to include "CLUSTERNAME" somewhere in the affinity group name. That was not clear to me. How is the destroyer going to clean up affinity groups with user-supplied names then?

Gal-Zaidman replied:
I think you are right on second thought, there is no harm in always adding the cluster name to a new affinity group that the installer creates, it will simplify the logic and improve the destroy logic :)
Change that part, thanks for the feedback!

staebler commented on pkg/types/ovirt/defaults/machinepools.go:
```diff
@@ -0,0 +1,42 @@
+// SetControlPlaneDefaults sets the defaults for the ControlPlane Machines.
+func SetControlPlaneDefaults(p *types.MachinePool) {
+	setMachinePool(p)
+	if p.Platform.Ovirt.AffinityGroups == nil {
```
```suggestion
	if len(p.Platform.Ovirt.AffinityGroups) == 0 {
```

Gal-Zaidman replied:
I'm ok with having 0 size, in that case no affinity groups will be created, that is the customization that will disable the affinity groups

staebler replied:
A roundtrip marshaling and unmarshaling of the install config is going to lose the difference between a missing `AffinityGroups` field and an `AffinityGroups` field set to an empty list.

Take for example an install-config.yaml with the following `controlPlane` section.
```
controlPlane:
  platform:
    ovirt:
      affinityGroups: []
```

Now, run `openshift-install create install-config`. The `controlPlane` section will become the following.
```
controlPlane:
  architecture: amd64
  hyperthreading: Enabled
  name: master
  platform:
    ovirt: {}
  replicas: 3
```

Now, run `openshift-install create install-config` again. The `controlPlane` section will become the following.
```
controlPlane:
  architecture: amd64
  hyperthreading: Enabled
  name: master
  platform:
    ovirt:
      affinityGroups:
      - name: CLUSTERNAME-controlplane
        priority: 5
      - name: CLUSTERNAME-cluster
        priority: 3
  replicas: 3
```

Gal-Zaidman replied:
Ok I see I guess we should handle that case, and when a user enters an empty array we shouldn't change the yaml and keep the empty array.
Great catch

staebler commented on pkg/types/ovirt/defaults/machinepools.go:
```diff
@@ -0,0 +1,42 @@
+// SetComputeDefaults sets the defaults for the Compute Machines.
+func SetComputeDefaults(p *types.MachinePool) {
+	setMachinePool(p)
+	if p.Platform.Ovirt.AffinityGroups == nil {
```
```suggestion
	if len(p.Platform.Ovirt.AffinityGroups) == 0 {
```

Gal-Zaidman replied:
I'm ok with having 0 size, in that case no affinity groups will be created, that is the customization that will disable the affinity groups

staebler commented on pkg/tfvars/ovirt/ovirt.go:
```diff
@@ -71,6 +78,50 @@ func TFVars(
+			}
+		}
+	}
+	// If the affinity group exist on master list don't create it twice
```
How are we going to deal with the affinity group name used for both control plane and compute but using different priorities for each? Is that going to be caught by validation?

Gal-Zaidman replied:
I will add this to the validations

staebler commented on pkg/types/ovirt/machinepool.go:
```diff
@@ -43,6 +47,12 @@ type CPU struct {
+// AffinityGroup defines the affinity group that the installer will create
+type AffinityGroup struct {
+	Name     string `json:"name"`
+	Priority int    `json:"priority,omitempty"`
```
You should not have an `int` type that is optional. If you want it to be optional, then the type should be `*int`. Are you sure that you even want it to be optional? I don't see code anywhere that sets the default priority.

Gal-Zaidman replied:
You are right, if a user decided that he wants to use customizations for affinity groups he should probably add all the fields for the Affinity Group, which will force him to think about it

staebler commented on pkg/types/ovirt/machinepool.go:
```diff
@@ -90,4 +100,8 @@ func (p *MachinePool) Set(required *MachinePool) {
 		p.OSDisk = required.OSDisk
 	}
+
+	if required.AffinityGroups != nil {
```
```suggestion
	if len(required.AffinityGroups) > 0 {
```

Gal-Zaidman commented on data/data/ovirt/affinity_group/main.tf:
```diff
@@ -0,0 +1,18 @@
+  vm_list      = var.masters_ids
+}
+
+resource "ovirt_affinity_group" "cluster_affinity_group" {
```
Initially this affinity group should've contain all the machines in the cluster, but we decided to make some changes to the affinity group layout and now it will be just for the compute nodes so I will change it :)

Gal-Zaidman commented on data/data/ovirt/affinity_group/variables.tf:
```diff
@@ -0,0 +1,22 @@
+
+variable "ovirt_control_plane_affinity_groups" {
+  type        = list(map(string))
+  description = "Control plane of the masters affinity groups that will be created."
```
Done thanks :)

Gal-Zaidman commented on data/data/ovirt/affinity_group/variables.tf:
```diff
@@ -0,0 +1,22 @@
+
+variable "masters_ids" {
+  type        = list(string)
+  description = "Name of the Affinity Group that will be created."
```
Fix that to describe the resource thanks for catching that!

Gal-Zaidman commented on pkg/asset/cluster/tfvars.go:
```diff
@@ -494,6 +494,10 @@ func (t *TerraformVariables) Generate(parents asset.Parents) error {
 			return err
 		}
 
+		workersInstallConfig := make([]*ovirt.MachinePool, 0)
```
+1

Gal-Zaidman commented on pkg/asset/machines/ovirt/machines.go:
```diff
@@ -73,6 +75,13 @@ func provider(platform *ovirt.Platform, pool *types.MachinePool, userDataSecret
 		MemoryMB:          pool.Platform.Ovirt.MemoryMB,
 		VMType:            string(pool.Platform.Ovirt.VMType),
 	}
+	ags := make([]string, 0)
```
+1

janoszen commented:
@Gal-Zaidman the blocking bug is now merged.

janoszen commented:
/retest

janoszen commented:
/test e2e-ovirt

Gal-Zaidman commented:
/test e2e-ovirt

staebler commented on pkg/asset/installconfig/ovirt/validation.go:
```diff
@@ -63,6 +66,129 @@ func validateMachinePool(con *ovirtsdk.Connection, child *field.Path, pool *ovir
+
+	clusterName, err := GetClusterName(con, ic.Ovirt.ClusterID)
+	if err != nil {
+		allErrs = append(allErrs, field.InternalError(fldPath, err))
```
Do you want to return here? The rest of the validation is worthless if we don't have a cluster name.

Gal-Zaidman replied:
Yes you are correct, fixed

staebler commented on pkg/asset/installconfig/ovirt/validation.go:
```diff
@@ -63,6 +66,129 @@ func validateMachinePool(con *ovirtsdk.Connection, child *field.Path, pool *ovir
+	}
+	hosts, err := FindHostsInCluster(con, clusterName)
+	if err != nil {
+		allErrs = append(allErrs, field.InternalError(fldPath, err))
```
Return here, too?

staebler commented on pkg/asset/installconfig/ovirt/validation.go:
```diff
@@ -63,6 +66,129 @@ func validateMachinePool(con *ovirtsdk.Connection, child *field.Path, pool *ovir
+		allErrs = append(allErrs, field.InternalError(fldPath, err))
+	}
+	for _, ag := range ic.Ovirt.AffinityGroups {
+		if _, found := mAgReplicas[ag.Name]; found {
```
Is there validation somewhere to ensure that the affinity group selected for a machine pool either is included in the affinity groups created by the installer or already exists in the oVirt cluster?

Gal-Zaidman replied:
Added one

staebler commented on pkg/types/ovirt/defaults/machinepools.go:
```diff
@@ -0,0 +1,34 @@
+// SetControlPlaneDefaults sets the defaults for the ControlPlane Machines.
+func SetControlPlaneDefaults(p *ovirt.Platform, mp *types.MachinePool) {
+	setMachinePool(mp)
+	setDefaultAffinityGroups(p, mp, DefaultControlPlaneAffinityGroupName)
```
It is surprising behavior to me that the afinity group of a machine pool gets set if and only if there is an affinity group specified with an exact name. At a minimum, this needs to be well documented in the `affinityGroups` field description in the machine pool type.

Gal-Zaidman replied:
Well I'm not sure I understood you, but this is only for the defaults.
The user can specify anything he wants basically

staebler commented on pkg/asset/machines/ovirt/machines.go:
```diff
@@ -9,10 +9,9 @@ import (
 	metav1 "k8s.io/apimachinery/pkg/apis/meta/v1"
 	"k8s.io/apimachinery/pkg/runtime"
 
+	ovirtprovider "github.com/openshift/cluster-api-provider-ovirt/pkg/apis/ovirtprovider/v1beta1"
```
Move this import into the previous section, with other non-installer imports.

Gal-Zaidman replied:
Done

staebler commented on pkg/types/ovirt/defaults/machinepools.go:
```diff
@@ -0,0 +1,34 @@
+}
+
+func setDefaultAffinityGroups(p *ovirt.Platform, mp *types.MachinePool, agName string) {
+	if mp.Platform.Ovirt.AffinityGroupsNames == nil {
```
This field is `omitempty`, so we need to treat nil and an empty slice identically.
```suggestion
	if len(mp.Platform.Ovirt.AffinityGroupsNames) == 0 {
```

Gal-Zaidman replied:
Done

staebler commented on pkg/tfvars/ovirt/ovirt.go:
```diff
@@ -2,12 +2,14 @@
-	"encoding/json"
+	"fmt"
 
+	"encoding/json"
```
Keep this import where it was.

Gal-Zaidman replied:
Done

staebler commented on pkg/asset/machines/ovirt/machines.go:
```diff
@@ -73,6 +72,20 @@ func provider(platform *ovirt.Platform, pool *types.MachinePool, userDataSecret
+	for _, ag := range platform.AffinityGroups {
+		uniqueNewAG[ag.Name] = ag
+	}
+	ags := make([]string, 0, len(pool.Platform.Ovirt.AffinityGroupsNames))
+	for _, agName := range pool.Platform.Ovirt.AffinityGroupsNames {
+		if _, ok := uniqueNewAG[agName]; ok {
+			// add the cluster name only if the affinity group is created by the installer
+			ags = append(ags, clusterID+"-"+agName)
+		} else {
+			ags = append(ags, agName)
+		}
+	}
```
```suggestion

	ags := make([]string, len(pool.Platform.Ovirt.AffinityGroupsNames))
	for i, agName := range pool.Platform.Ovirt.AffinityGroupsNames {
		if _, ok := uniqueNewAG[agName]; ok {
			// add the cluster name only if the affinity group is created by the installer
			ags[i] = clusterID+"-"+agName
		} else {
			ags[i] = agName
		}
	}
```

Gal-Zaidman replied:
Done

staebler commented on pkg/tfvars/ovirt/ovirt.go:
```diff
@@ -73,6 +79,21 @@ func TFVars(
 	return json.MarshalIndent(cfg, "", "  ")
 }
+
+func handleAffinityGroups(ags []*ovirt.AffinityGroup, infraID string) []map[string]interface{} {
```
This is not a very descriptive name for this function.

staebler commented on pkg/tfvars/ovirt/ovirt.go:
```diff
@@ -73,6 +79,21 @@ func TFVars(
 }
+
+func handleAffinityGroups(ags []*ovirt.AffinityGroup, infraID string) []map[string]interface{} {
+	var agMap []map[string]interface{}
+	for _, ag := range ags {
+		agMap = append(agMap, map[string]interface{}{
+			"name":        fmt.Sprintf("%s-%s", infraID, ag.Name),
+			"priority":    float64(ag.Priority),
+			"description": ag.Description,
+			"enforcing":   ag.Enforcing,
+		})
+	}
```
```suggestion
	agMap := make([]map[string]interface{}, len(ags))
	for i, ag := range ags {
		agMap[i] = map[string]interface{}{
			"name":        fmt.Sprintf("%s-%s", infraID, ag.Name),
			"priority":    float64(ag.Priority),
			"description": ag.Description,
			"enforcing":   ag.Enforcing,
		}
	}
```

Gal-Zaidman replied:
Done

staebler commented on pkg/tfvars/ovirt/ovirt.go:
```diff
@@ -73,6 +79,21 @@ func TFVars(
 }
+
+func handleAffinityGroups(ags []*ovirt.AffinityGroup, infraID string) []map[string]interface{} {
+	var agMap []map[string]interface{}
```
This is a confusing name for this variable, since it is not a map but is instead a slice.

Gal-Zaidman replied:
Done

staebler commented on pkg/tfvars/ovirt/ovirt.go:
```diff
@@ -73,6 +79,21 @@ func TFVars(
+	for _, ag := range ags {
+		agMap = append(agMap, map[string]interface{}{
+			"name":        fmt.Sprintf("%s-%s", infraID, ag.Name),
+			"priority":    float64(ag.Priority),
```
Why does this need to be converted to a float? Is it necessary for the way that the value gets marshaled?

Gal-Zaidman replied:
Yes the oVirt API expects a float so the terraform provider expects a float

staebler replied:
Terraform does not distinguish between whole numbers and fractional values.

staebler commented on pkg/destroy/ovirt/destroyer.go:
```diff
@@ -144,6 +148,25 @@ func (uninstaller *ClusterUninstaller) removeTemplate(con *ovirtsdk.Connection)
+		return err
+	}
+	for _, ag := range res.MustGroups().Slice() {
+		if strings.Contains(ag.MustName(), uninstaller.Metadata.InfraID) {
```
Don't we only want to remove the affinity group if it has the prefix of "{INFRA_ID}-"?
```suggestion
		if strings.HasPrefix(ag.MustName(), fmt.Sprintf("%s-", uninstaller.Metadata.InfraID)) {
```

Gal-Zaidman replied:
Well, it is a unique string so contains felt safe enough but sure.
Done

staebler commented on pkg/destroy/ovirt/destroyer.go:
```diff
@@ -144,6 +148,25 @@ func (uninstaller *ClusterUninstaller) removeTemplate(con *ovirtsdk.Connection)
+			uninstaller.Logger.Infof("Removing AffinityGroup %s", ag.MustName())
+			_, err := affinityGroupService.GroupService(ag.MustId()).Remove().Send()
+			if err != nil {
+				uninstaller.Logger.Errorf("failed to remove AffinityGroup: %s", err)
```
Don't we also want to ultimately return an error from `removeAffinityGroups` when we fail to remove an affinity group, so that we keep trying?

But I see that we are not retrying for other resources. I think that is a serious deficiency that needs to be addressed for the ovirt destroyer, in general.

Gal-Zaidman replied:
Opened a new bug for that https://bugzilla.redhat.com/show_bug.cgi?id=1947424 Thanks!

Gal-Zaidman commented on pkg/types/ovirt/platform.go:
```diff
@@ -33,4 +33,16 @@ type Platform struct {
+	Name        string `json:"name"`
+	Priority    int    `json:"priority"`
+	Description string `json:"description"`
+	Enforcing   bool   `json:"enforcing"`
```
I think I prefer keeping it a bool since the ovirt API expects a bool, and if we will just an enum is will just be deferent then ovirt plus make the code a bit harder to read :/

Gal-Zaidman commented on pkg/types/ovirt/platform.go:
```diff
@@ -33,4 +33,16 @@ type Platform struct {
+
+// AffinityGroup defines the affinity group that the installer will create
+type AffinityGroup struct {
+	Name        string `json:"name"`
```
Done

Gal-Zaidman commented:
/test e2e-ovirt

Gal-Zaidman commented on pkg/asset/installconfig/ovirt/validation.go:
```diff
@@ -63,6 +66,129 @@ func validateMachinePool(con *ovirtsdk.Connection, child *field.Path, pool *ovir
+	return allErrs
+}
+
+func validateAffinityGroupFields(platform ovirt.Platform, fldPath *field.Path) field.ErrorList {
```
Added optional on description, and since enforcing is its zero value is false so no need for validation on it

Gal-Zaidman commented on pkg/asset/installconfig/ovirt/validation.go:
```diff
@@ -63,6 +66,129 @@ func validateMachinePool(con *ovirtsdk.Connection, child *field.Path, pool *ovir
+		if ag.Name == "" {
+			allErrs = append(
+				allErrs,
+				field.Invalid(fldPath, ag,
```
Done

Gal-Zaidman commented on pkg/asset/installconfig/ovirt/validation.go:
```diff
@@ -63,6 +66,129 @@ func validateMachinePool(con *ovirtsdk.Connection, child *field.Path, pool *ovir
+func validateAffinityGroupFields(platform ovirt.Platform, fldPath *field.Path) field.ErrorList {
+	allErrs := field.ErrorList{}
+	for _, ag := range platform.AffinityGroups {
+		if ag.Name == "" {
```
Done

Gal-Zaidman commented:
/test e2e-ovirt

Gal-Zaidman commented:
/test e2e-ovirt

Gal-Zaidman commented:
/test e2e-ovirt

Gal-Zaidman commented:
/test e2e-ovirt

Gal-Zaidman commented:
/test e2e-ovirt

Gal-Zaidman commented:
/retest

Gal-Zaidman commented:
/retest

Gal-Zaidman commented:
/retest

Gal-Zaidman commented:
/test e2e-ovirt

Gal-Zaidman commented:
@staebler sorry this took a while to be ready, we needed to make some changes to the terraform provider to do this effort right.
Can you please re-review this PR, there are no real changes from your last review beside the fixes you requested :)

janoszen commented:
/retest

Gal-Zaidman commented:
/unhold

Gal-Zaidman commented:
/retest

Gal-Zaidman commented:
/retest

janoszen commented:
/retest

janoszen commented:
/retest

janoszen commented:
/retest

staebler commented on pkg/types/ovirt/platform.go:
```diff
@@ -33,4 +33,16 @@ type Platform struct {
 	// +optional
 	DefaultMachinePlatform *MachinePool `json:"defaultMachinePlatform,omitempty"`
+
+	// AffinityGroupsNames contains the names of the RHV affinity groups that the machines will join
```
This is not the correct name of the field.
Field descriptions should end in periods.
```
	// AffinityGroups is the list RHV affinity groups that the machines will join.
```

Gal-Zaidman replied:
Done

staebler commented on pkg/types/ovirt/defaults/platform.go:
```diff
@@ -7,9 +7,39 @@ import (
 // DefaultNetworkName is the default network name to use in a cluster
 const DefaultNetworkName = "ovirtmgmt"
 
+// DefaultControlPlaneAffinityGroupName is the default affinity group name for the control plain VMs
```
```suggestion
// DefaultControlPlaneAffinityGroupName is the default affinity group name for the control plane VMs
```

Gal-Zaidman replied:
Done

staebler commented on pkg/types/ovirt/defaults/platform.go:
```diff
@@ -7,9 +7,39 @@ import (
+	return &ovirt.AffinityGroup{
+		Name:        DefaultControlPlaneAffinityGroupName,
+		Priority:    5,
+		Description: "AffinityGroup for spreading each control plane machines to a different host",
```
```suggestion
		Description: "AffinityGroup for spreading each control plane machine to a different host",
```

Gal-Zaidman replied:
Done

staebler commented on pkg/types/ovirt/platform.go:
```diff
@@ -33,4 +33,16 @@ type Platform struct {
+
+	// AffinityGroupsNames contains the names of the RHV affinity groups that the machines will join
+	// +optional
+	AffinityGroups []*AffinityGroup `json:"affinityGroups"`
```
This does not need to be a slice of pointers. The extra indirection adds unnecessary complexity.
```
	AffinityGroups []AffinityGroup `json:"affinityGroups"`
```

Gal-Zaidman replied:
Done

staebler commented on pkg/types/ovirt/platform.go:
```diff
@@ -41,8 +41,14 @@ type Platform struct {
+	Priority int `json:"priority"`
+	// Description of the affinity group
+	// +optional
 	Description string `json:"description"`
```
```suggestion
	Description string `json:"description,omitempty"`
```

Gal-Zaidman replied:
Done

staebler commented on pkg/asset/installconfig/ovirt/validation.go:
```diff
@@ -2,9 +2,9 @@ package ovirt
 
 import (
 	"fmt"
-
```
This blank line still needs to be put back.

Gal-Zaidman replied:
Done

staebler reviewed (commented):
What is the plan for the commits? Are you planning to squash them down? If not, then the commit messages need some work. At a minimum they should make it clear that the commit is specific to oVirt. The commit messages should each stand on their own and not require the context of the entire PR to make it clear what is being done in the commit.
See https://github.com/openshift/installer/blob/master/CONTRIBUTING.md#commit-message-format

Gal-Zaidman commented:
> What is the plan for the commits? Are you planning to squash them down? If not, then the commit messages need some work. At a minimum they should make it clear that the commit is specific to oVirt. The commit messages should each stand on their own and not require the context of the entire PR to make it clear what is being done in the commit.
> See https://github.com/openshift/installer/blob/master/CONTRIBUTING.md#commit-message-format

I squashed some commits together and added description for each commit

Gal-Zaidman commented:
/retest

Gal-Zaidman commented on pkg/asset/installconfig/ovirt/validation.go:
```diff
@@ -63,6 +66,129 @@ func validateMachinePool(con *ovirtsdk.Connection, child *field.Path, pool *ovir
+				fldPath,
+				errors.Errorf("failed listing affinity groups for cluster %v", platform.ClusterID)))
+	}
+	for _, ag := range res.MustGroups().Slice() {
```
Yes it would, thanks for the catch

Gal-Zaidman commented:
/retest

Gal-Zaidman commented:
/retest

janoszen commented:
/retest

janoszen commented on data/data/ovirt/affinity_group/main.tf:
```diff
@@ -0,0 +1,9 @@
+resource "ovirt_affinity_group" "affinity_groups" {
```
Shouldn't the `vm_list` be put in the `ignore_changes` lifecycle attribute here? After all, the PR that removes this attribute is not merged and there are some valid discussions around this.

Gal-Zaidman replied:
Well, we are not passing it so it is never changed.
The problem in the past was the dependency that we had between the bootstrap and masters that caused the affinity groups to get reevaluated on destroy but we removed that thanks to moving the lock to the affinity group function in the VM resource, so no problem now.

Gal-Zaidman replied:
Done

janoszen commented on data/data/ovirt/affinity_group/outputs.tf:
```diff
@@ -0,0 +1,3 @@
+output "ovirt_affinity_group" {
```
Nitpick: this could be named `ovirt_affinity_group_count`

Gal-Zaidman replied:
Done

janoszen commented on pkg/asset/installconfig/ovirt/client.go:
```diff
@@ -69,3 +69,30 @@ func FetchVNICProfileByClusterNetwork(con *ovirtsdk.Connection, clusterID string
+func GetClusterName(con *ovirtsdk.Connection, clusterID string) (string, error) {
+	clusterResponse, err := con.SystemService().ClustersService().ClusterService(clusterID).Get().Send()
+	if err != nil {
+		return "", err
```
Please add context to the error here, e.g. `fmt.Errorf("failed to fetch cluster %s (%w)", clusterID, err)`

Gal-Zaidman replied:
Done

janoszen commented on pkg/asset/installconfig/ovirt/client.go:
```diff
@@ -69,3 +69,30 @@ func FetchVNICProfileByClusterNetwork(con *ovirtsdk.Connection, clusterID string
+	res, err := con.SystemService().HostsService().
+		List().Search(fmt.Sprintf("cluster=%s", cName)).Send()
+	if err != nil {
+		return nil, err
```
Please add context to the error here, like above.

Gal-Zaidman replied:
Done

janoszen commented on pkg/types/ovirt/machinepool.go:
```diff
@@ -24,6 +24,11 @@ type MachinePool struct {
 	// +optional
 	VMType VMType `json:"vmType,omitempty"`
+
+	// AffinityGroupsNames contains the name of the OpenShift cluster affinity groups
```
Suggestion:

> AffinityGroupNames contains a list of oVirt affinity group names that will be used to separate the OpenShift cluster nodes from each other to separate physical machines. If the affinity group is not created in the install config it is assumed that the AG already exists in oVirt.

Gal-Zaidman replied:
Changed it to:
```
AffinityGroupsNames contains a list of oVirt affinity group names that the newly created machines will join.
The affinity groups should exist on the oVirt cluster or created by the OpenShift installer.
```
I don't want to say for what the affinity group will be used for :)

janoszen commented on pkg/types/ovirt/platform.go:
```diff
@@ -33,4 +33,22 @@ type Platform struct {
 	// +optional
 	DefaultMachinePlatform *MachinePool `json:"defaultMachinePlatform,omitempty"`
+
+	// AffinityGroups contains the RHV affinity groups that the installer will create.
```
Suggestion:

> AffinityGroups contains a list of affinity groups that will be created by the installer. If an affinity group is not listed here, but specified in AffinityGroupNames it is assumed that the affinity group already exists in oVirt.

Gal-Zaidman replied:
I don't want to add a reference to the machine field here, I think that saying it on the machine field is more than enough

Gal-Zaidman commented:
/retest

janoszen commented:
/retest

Gal-Zaidman commented:
/retest

janoszen commented:
/retest

Gal-Zaidman commented:
/retest

Gal-Zaidman commented:
/retest

janoszen commented:
/lgtm

Gal-Zaidman commented:
/retest

Gal-Zaidman commented:
/retest

Gal-Zaidman commented:
/retest

Gal-Zaidman commented:
/retest

Gal-Zaidman commented:
/retest

Gal-Zaidman commented:
/retest

Gal-Zaidman commented:
/retest

staebler commented:
/approve
